bigquery/google/cloud/bigquery/client.py

```diff
@@ -1129,13 +1178,33 @@ def query_rows(self, query, job_config=None, job_id=None, timeout=None,
             from the total number of rows in the current page:
             ``iterator.page.num_items``).
 
-        :raises: :class:`~google.cloud.exceptions.GoogleCloudError` if the job
-            failed or  :class:`TimeoutError` if the job did not complete in the
-            given timeout.
+        :raises:
+            :class:`~google.api.core.exceptions.GoogleAPICallError` if the
+            job failed or :class:`concurrent.futures.TimeoutError` if the job
+            did not complete in the given timeout.
+
+            Exceptions thrown by this function will be ammended with
+            ``project``, ``job_id``, and ``full_job_id`` string properties.
+            Use :meth:`get_job` with the ``project`` and ``job_id``
+            properties to get more information about the query job and wait
+            for it to finish or use :meth:`cancel_job` to cancel it.
+
+            Use the ``full_job_id`` (of the format ``project:job_id``) when
+            communicating with support about a failed job.
         """
-        job = self.query(
-            query, job_config=job_config, job_id=job_id, retry=retry)
-        return job.result(timeout=timeout)
+        job_id = _make_job_id(job_id, job_id_prefix)
+
+        try:
+            job = self.query(
+                query, job_config=job_config, job_id=job_id, retry=retry)
+            rows_iterator = job.result(timeout=timeout)
+        except (GoogleAPICallError, concurrent.futures.TimeoutError) as exc:
+            exc.project = self.project
```

Comment:
Can you tell me why this information is needed?

Reply:
If a call to `query_rows()` fails and someone wants to investigate further, they will need the full job ID (including the project). This is the first information that support or BigQuery engineers ask for.

If a call to query_rows() times out, the query is still running in the background. In this case, the user will either want to get more information about the job or cancel it.

Reply:
So I'm kind of -1 on this.

1. It adds new properties to an object, which is generally frowned upon. There's no docstrings for this, there's no declaration of the interface. Finally, if we were to add `__slots___` to our exception classes, this would fail outright.
2. It doesn't cause any behavioral change in the exceptions. It adds the info, sure, but it doesn't print this info out in `str`.

@dhermes what do you think?

Reply:
Other options:

- Make new exception types to wrap this just for `query_rows()`? That seems even yuckier than adding properties.
- Change the behavior to call cancel on the job in case of error or timeout. (Noting in the docs that cancel is best effort.)

Calling `cancel` is less flexible. Then again, someone who is using `query_rows()` instead of `query()`is actually asking for less flexibility. It also does not address that fact that they don't have the job ID, but they could potentially call `list_jobs` to figure it out.

Reply:
I'm going with "Change the behavior to call cancel on the job in case of error or timeout. (Noting in the docs that cancel is best effort.)" Will open a second PR as bigquery-b2 is now merged to master.